### ISSUE TITLE
🟪 Contest Worker Policy

### ISSUE THREAD

**Issue by Rotorsoft** [hicommonwealth/commonwealth#6822]
# Description

This feature entails the creation of a new integration policy designed to respond to platform events, specifically content creation and upvoting. The primary objective is to synchronize these events into the on-chain contest manager. This synchronization process will involve invoking protocol domain services responsible for adding content and votes to the contest manager.

## Stakeholders

| Product   | Engineering |
| --------- | ----------- |
| @zakhap      | @Rotorsoft        |

## Acceptance Criteria

- The new integration policy effectively captures platform events related to content creation and upvoting.
- Synchronization of events into the on-chain contest manager occurs without errors or data loss
- Protocol domain services for adding content and votes to the contest manager are invoked as intended.
- Data consistency and accuracy are maintained between platform events and the on-chain contest manager.

## Model
[Contest Worker Policy](https://miro.com/app/board/uXjVNZ8oB3U=/?moveToWidget=3458764573734387460&cot=14)

## Engineering Requirements

Integration
  - Use existing transactional outbox pattern

Routing
- `ThreadCreated` event params:
  - Thread creator address
  - Thread page URL
- `ThreadUpvoted` event params:
  - Voter address
  - Thread ID

References
  - Requires the `All Contests` projection which should contain the contract address

Business Rules
  - @rbennettcw Contests can be paused/resumed, so this policy should only act on running contests - #7471 
  - Policy: `Contest Worker Policy`
  - Invokes: `Add Content` and `Add Vote` commands on Contest Manager
  - Command schemas:
    - `Add Content`: Contract address, creator address, content page URL
    - `Add Vote`: Contract address, voter address, content ID

Response
  - If the commands invoked on Contest Manager fail, then log the errors to rollbar (setup high priority alerting?) and attempt to retry with exponential backoff.
    - When the policy returns (void) without exceptions, the queue is Ack-ed
    - When the policy throws a retry-able error (like service unreachable), the queue can retry (we can define multiple retry strategies later if needed) - after max retries, msg is sent to DLQ
    - When the policy throws a fatal error, the queue gets blocked and the errors logged in rollbar.
  - Possible error types
    - RPC connection failure
    - Transaction failure (e.g. not enough gas)
    - Contract error (e.g. contest ended)?

**Comment by Rotorsoft** [comment]
@rbennettcw let's start refining the engineering requirements and provide a rough estimate

**Comment by ianrowan** [comment]
I created a duplicate of this Ticket from the protocol perspective as I wasn't sure if it would be handled by Proto or Platform. Regardless info in the ticket should be helpful: https://github.com/hicommonwealth/commonwealth/issues/6989

feel free to close my ticket if not needed/just use it as a reference 

CC @Rotorsoft @rbennettcw @ForestMars @jnaviask

**Comment by jnaviask** [comment]
@ianrowan @Rotorsoft @rbennettcw would love for protocol team to coordinate on hooks/helpers for protocol-related functionality as needed for platform team to put together the actual policy.

**Comment by ianrowan** [comment]
@Rotorsoft @rbennettcw Once work begins on this ticket, would be great to sync on the arch decisions to help inform #7052

**Comment by Rotorsoft** [comment]
From the architecture POV:
- Protocol utilities are considered "Domain Services" (in DDD terms). 
- We should locate all protocol domain services within a lib folder (currently `/libs/model/services`. 

Just for context:
A domain service is a type of service that operates on domain objects and logic. It encapsulates operations and behaviors that don't naturally fit within the boundaries of a single entity or value object -> stateless, can operate on multiple entities.

**Comment by jnaviask** [comment]
Cannot be completed until #7052 is unblocked, but @ianrowan will help with stubbing out interfaces?

**Comment by ianrowan** [comment]
I've stubbed the helpers in 7052-contests-policy-contract-helpers-stubs](https://github.com/hicommonwealth/commonwealth/tree/7052-contests-policy-contract-helpers-stubs)

Should allow dev to start on this whenever ready - let me know as development starts if any modificaiton/additional helpers are required and I'll add to branch

**Comment by rbennettcw** [comment]
The Contest Worker Policy will respond to the Thread Created and Thread Upvoted events, so #7240 and #7252 are the relevant refactor tickets.

The scope of this ticket will focus on consuming the aforementioned events and invoking protocol commands from Ian's branch– functions `addContent` and `voteContent`.

**Comment by Rotorsoft** [comment]
@rbennettcw do we have enough information to update the ticket description now?

**Comment by rbennettcw** [comment]
@Rotorsoft  After chatting with Ian, I ran across some details that may require a change to the model?

In order for the policy to invoke the `addContent` and `voteContent` protocol functions, the policy needs the contest contract address:
  - Source of truth is onchain
    - it's first accessible on the frontend after the user deploys a Contests contract. Getting the address to the backend can be done via push (API request), which is a somewhat fragile solution, or pull via some blockchain listener integration (Tim's notification system can possibly handle this?).
    - Push would be easiest to implement.
  - We need to create the projection (to be stored in the DB) for contest metadata which would include the contract address. Then the policy will have all the data required to invoke the Contest Manager commands.
    - Additionally, content is linked to contests by Topic (similarly to how gated actions only affects actions within linked topics), so perhaps the contest metadata projection should store linked topics too.

**Comment by Rotorsoft** [comment]
@rbennettcw do you see the "All contests" projection on top of the policy (in the model)? That's where you will find all the details to invoke the protocol service

**Comment by rbennettcw** [comment]
@Rotorsoft Ok great, model looks sufficient then!

**Comment by rbennettcw** [comment]
Draft of engineering requirements:

# Contests

Integration
  - Use existing transactional outbox pattern

Routing
- `ThreadCreated` event params:
  - Thread creator address
  - Thread page URL
- `ThreadUpvoted` event params:
  - Voter address
  - Thread ID

References
  - Requires the `All Contests` projection which should contain the contract address

Business Rules
  - Policy: `Contest Worker Policy`
  - Invokes: `Add Content` and `Add Vote` commands on Contest Manager
  - Command schemas:
    - `Add Content`: Contract address, creator address, content page URL
    - `Add Vote`: Contract address, voter address, content ID

Response
  - If the commands invoked on Contest Manager fail, then log the errors to rollbar (setup high priority alerting?) and attempt to retry with exponential backoff.
    - Not sure about specific implementation details of how the notification system can/will handle this, but retry logic would need to be queue-based to work between app deployments.
  - Possible error types
    - RPC connection failure
    - Transaction failure (e.g. not enough gas)
    - Contract error (e.g. contest ended)?

**Comment by Rotorsoft** [comment]
@rbennettcw you can update the ticket description so we can always have the latest requirements on top... tracking changes in comments is not fun (at least for me 😆) 

On the response section, there are a few patterns we can use, and the consumer utility that @timolegros created should be able to manage these scenarios (we need to define conventions here, but something like):
- When the policy returns (void) without exceptions, the queue is Ack-ed
- When the policy throws a retry-able error (like service unreachable), the queue can retry (we can define multiple retry strategies later if needed) - after max retries, msg is sent to DLQ
- When the policy throws a fatal error, the queue gets blocked and the errors logged in rollbar.
- There might be other scenarios, let's define them and the error types.

**Comment by rbennettcw** [comment]
This will be unblocked by #7593
